### ISSUE TITLE
VZD: Update Locations

### ISSUE THREAD

**Issue by sergiogcx** [cityofaustin/atd-data-tech#4354]
Once the Level 5 street levels are inserted in the database, a few things need to be updated:

1. Make sure VZE only shows level 4 and below, within the VZE configuration this should be an easy fix.

2. Update ETL: https://github.com/cityofaustin/atd-airflow/blob/master/dags/python_scripts/atd_vzd_update_cr3_locations.py

3. Update AWS triggers if needed.

**Comment by frankhereford** [comment]
Jace and I spoke this afternoon, and suggested that it would be helpful if I were to memorialize my understanding of the process around the impacts and methods for adding these new polygons to the atd_txdot_locations table. 

Currently, the atd_txdot_locations contains the polygon set derived from the ASMP streets level 1-4. These polygons are created to never overlap one another, so that a crash can potentially fall into either a single polygon or none at all. The set of crashes which are allowed to fall into these polygons is defined as the super-set of crashes minus the set crashes which fall along the "main-lanes", as defined by a pair of shapefiles originally located by Joel Meyer during the polygon creation process. This main-lane exclusion gives us a route to differentiate crashes which happen below an overpass on a level 5 street from those which happen on the surface street crossing over a level 5, such as the Martin Luther King Jr Blvd. bridge over IH 35. 

While level 5 polygons never overlap with other level 5 polygons, they _can_ overlap with level 1-4 polygons. This introduces a winkle where associating crashes to a polygon can no longer be performed with a spatial join between the one of the two crashes tables and the locations table because if a crash occurs in an area a level 5 polygon overlaps a level 1-4, (such as an overpass type situation described above), the result set of that spatial join would contain two records, one for each polygon the crash is contained by, and this does not represent the actual crash event.

So, the proposed solution which has been discussed and developed is a set of SQL files as follows:

1. A file which contains a set of UPDATEs to various 1-4 polygons to update their geometry based on feedback from VZE users. It also contains a set of DELETE statements to remove some polygons which were generated from ASMP streets that were not actually representative of roads in on the ground. And finally, a set of INSERTs which create the level 5 polygons. All of these operations are on the atd_txdot_locations table. 

2. A file which contains CREATE MATERIALIZED VIEW instructions which build materialized views that define the various sets of crashes discussed above as well, as well as a materialized view which joins the appropriate set of crashes to the appropriate set of polygons to allow for a crashes polygon to be queried directly. 

3. A file which contains two queries that can be used to re-associate all crashes to the correct polygon, meaning to re-associate crashes which may have changed their polygon due to the UPDATEs in step 1, or because the crash now has a polygon (level 5) where previously it did not have one to fall into. These queries are intended to be used once only during the process where these new polygons and new changes are introduced. 

4. A file which contains modifications to certain functions that are exposed by Hasura (and by extension, the VZV/VZE application) to limit the application so that it only sees level 1-4 polygons. The intent here is to minimize the amount of work which needs to be done to the VZV/VZE to maintain its current functionality, as it was developed prior to the introduction of the level 5 polygons.

5. A file which contains the REFRESH MATERIALIZED VIEW commands for the materialized views created in step 2 to be incorporated into the nightly ETL job which loads new crashes. 

Sergio and I discussed this proposed solution and my understanding is that it has tentative approval as a way forward, pending testing in staging. One issue yet to be fully resolved includes what happens if a crash has its location updated -- does it need to be re-associated immediately or it it acceptable to have it re-associated by a scheduled job, such as the ETL process? In the case of the prior, modifications to the lambda function which does the polygon-crash association may be required to describe the association logic where a crash first tries to bind to a 1-4 polygon and failing that, tries to bind to a level 5 polygon. 

Currently, these files operate on the "polygons" table, but I have since learned that it is actually the atd_txdot_locations table which is used by the VZE/VZV application. I am refactoring the programs which generate these SQL files to operate on the correct table, including checking that any column name changes are reflected correctly.
